feat: add custom cursor interface on windows

Kingtous opened the pull request:
This PR adds an additional interface to set custom cursor on windows, which fixes issues mentioned below.

Currently,  Customizing `MouseCursor` on windows is not working, the customed mouse will restore to arrow once it moves. See https://github.com/flutter/flutter/issues/31952#issuecomment-1056115014 
https://github.com/flutter/flutter/issues/99484
for details. Variable `current_cursor_` in `cursor handler` will not update when we set our custom cursors, and we cannot change this field to prevent the cursor from restoring due to lacking of interface.

The reason is that cursor handler on windows has no interface to change current cursor to custom cursor, only support setting windows self-contained icons.  

Changes:
- add new function handled by method channel of `SystemCursors.mousecursors`, which can change `current_cursor_`, make possible to handle custom cursor on windows correctly.

![Screenshot_20220914_183105](https://user-images.githubusercontent.com/39793325/190131253-8ee3d739-ec64-43b8-9ac3-70d507df34c3.png)

We have successfully tested this interface in https://github.com/Kingtous/rustdesk_flutter_custom_cursor/blob/8f3be4ac68b4fb20d05b8d30df4c13d8fd77f9c4/lib/flutter_custom_cursor.dart#L113, which works perfectly on windows.

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

loic-sharma commented on shell/platform/windows/cursor_handler.cc:
```diff
@@ -45,6 +46,31 @@ void CursorHandler::HandleMethodCall(
+    auto width = std::get<int>(map.at(flutter::EncodableValue("width")));
+    auto height = std::get<int>(map.at(flutter::EncodableValue("height")));
+    auto hotx = std::get<double>(map.at(flutter::EncodableValue("hotx")));
+    auto hoty = std::get<double>(map.at(flutter::EncodableValue("hoty")));
```
Please validate that the expected arguments are provided. Otherwise, return an error result

Kingtous replied:
Yes, I agree. Instead of getting the variable directly from std::get, we use `find` to check the variable existence.

loic-sharma commented on shell/platform/windows/cursor_handler.cc:
```diff
@@ -45,6 +46,31 @@ void CursorHandler::HandleMethodCall(
+    icon_info.yHotspot = hoty;
+    icon_info.hbmMask = bitmap;
+    icon_info.hbmColor = bitmap;
+    cursor = CreateIconIndirect(&icon_info);
```
It looks like [this can fail and return `NULL`](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-createiconindirect#return-value). Can we check this and return an error result?

Kingtous replied:
yes, i agree with this. a check is added in L108

loic-sharma commented on shell/platform/windows/cursor_handler.cc:
```diff
@@ -45,6 +46,31 @@ void CursorHandler::HandleMethodCall(
+    // Flutter returns rawRgba, which has 8bits*4channels
+    auto bitmap = CreateBitmap(width, height, 1, 32, &buffer[0]);
+    if (bitmap == nullptr) {
+      result->Error("Argument error", "Invalid rawRgba bitmap from flutter");
```
What do you think of this?

```suggestion
      result->Error("Argument error", "Argument buffer must contain valid rawRgba bitmap");
```

Kingtous replied:
I agree with that. :)

Kingtous commented on shell/platform/windows/cursor_handler.cc:
```diff
@@ -45,6 +51,65 @@ void CursorHandler::HandleMethodCall(
+    cursor = CreateIconIndirect(&icon_info);
+    DeleteObject(bitmap);
+    if (cursor == nullptr) {
+      result->Error("Argument error", "Create Icon failed, argument buffer must contain valid rawRgba bitmap");
```
CreateIconIndirect return NULL will be catched here.

loic-sharma commented:
@Kingtous Thanks for following up quickly! Could you add a unit test that sends a `setSystemCursor` message and uses `MockWindowBindingHandlerDelegate` to verify the cursor is set as expected? Once that's ready, we should get approval from @dkwingsmt as they worked in this space.

dkwingsmt commented on shell/platform/windows/cursor_handler.cc:
```diff
@@ -11,8 +11,14 @@
 static constexpr char kChannelName[] = "flutter/mousecursor";
 
 static constexpr char kActivateSystemCursorMethod[] = "activateSystemCursor";
+static constexpr char kSetSystemCursorMethod[] = "setSystemCursor";
 
 static constexpr char kKindKey[] = "kind";
+static constexpr char kCustomCursorBufferKey[] = "buffer";
+static constexpr char kCustomCursorHotxKey[] = "hotx";
+static constexpr char kCustomCursorHotyKey[] = "hoty";
+static constexpr char kCustomCursorWidthKey[] = "width";
+static constexpr char kCustomCursorHeightKey[] = "height";
```
```suggestion
static constexpr char kActivateSystemCursorMethod[] = "activateSystemCursor";
static constexpr char kKindKey[] = "kind";

// Also: Can you document this method here (the arguments and their types, and return format.)
static constexpr char kSetSystemCursorMethod[] = "setSystemCursor";
static constexpr char kCustomCursorBufferKey[] = "buffer";
static constexpr char kCustomCursorHotxKey[] = "hotx";
static constexpr char kCustomCursorHotyKey[] = "hoty";
static constexpr char kCustomCursorWidthKey[] = "width";
static constexpr char kCustomCursorHeightKey[] = "height";
```

dkwingsmt replied:
Also why is this new method called `set system cursor` while it's actually a custom cursor?

Kingtous replied:
Maybe "setCustomCursor" is more suitable for this method?

dkwingsmt commented:
Thank you. I didn't know you can create `ICONINFO` using a buffer.

However, I'm hesitant on merging the method channel, since we are hoping to design a universal API (method channel) for all platforms. But I also understand that we need to develop one step at a time. So I think there are two options:
1. This PR only includes embedding APIs (such as `SetFlutterCursor`), just enough for you to develop a plugin. The plugin will listen to the channel, create `ICONINFO`, and call the new embedding APIs.
2. Merge this PR, but we might deprecate these methods in the future. (When we come up with a universal API, we'll probably adapt the current legacy API onto the new one for a while for compatibility.)

loic-sharma commented on shell/platform/windows/cursor_handler.cc:
```diff
@@ -45,6 +51,79 @@ void CursorHandler::HandleMethodCall(
+          "Missing argument width while trying to customize system cursor");
+      return;
+    }
+    auto width = std::get<int>(arguments.at(flutter::EncodableValue("width")));
```
Could this use `width_iter->second` instead? Same comment for other arguments

Kingtous replied:
Sure

Kingtous commented:
> Thank you. I didn't know you can create `ICONINFO` using a buffer.
> 
> However, I'm hesitant on merging the method channel, since we are hoping to design a universal API (method channel) for all platforms, and the current method is likely platform dependent. But I also understand that we need to develop one step at a time. So I think there are two options:
> 
> 1. Make this PR only includes embedding APIs (such as `SetFlutterCursor`), just enough for you to develop a plugin. The plugin will listen to the channel, create `ICONINFO`, and call the new embedding APIs.
> 2. Merge this PR, but we might deprecate these methods in the future. (When we come up with a universal API, we'll probably adapt the current legacy API onto the new one for a while for compatibility.)
> 
> Also, the flow I'm imagining is, instead of setting the cursor using the buffer directly, which you have to do every time you set the same cursor,
> 
> 1. Create a cursor using the buffer, and assign it with a name (string)
> 2. Activate the cursor using the string.

Sounds great for a universal API to change cursor, which is a very important feature, especially for games. Although this PR is providing a new feature, it more like a  bugfix. It fills the limitation(or fix bugs?) with custom cursor on windows, currently.(by extending offcial `MouseCursor`, `MouseCursorSession` abstract class, linux/macOS works great)

This problem has blocked so many developers as mentioned in https://github.com/flutter/flutter/issues/31952#issuecomment-1056115014 https://github.com/flutter/flutter/issues/99484

I think the two options are all ok, and I prefer Option 2. Currently, this PR can be a temporary and fast workaround to uniform custom cursor behaviors on all platforms. Devs can change custom cursor easily by providing a raw RGBA buffer, without any win32 knowledge.

here are some questions:
Option 1: I am not very familiar with `embedding API`, can it be invoked without a `MethodChannel`?

I've thought about `custom cursor cache`, but it may bring a more complex custom cursor logic, which may not suitable for keeping this embedding API simple. Also, a cursor buffer is not so big, it only is invoked when the system cursor hovers on the specific item.

Kingtous commented:
> @Kingtous Thanks for following up quickly! Could you add a unit test that sends a `setSystemCursor` message and uses `MockWindowBindingHandler` to verify the cursor is set as expected? (If you want to go the extra mile, ideally `activateSystemCursor` would be tested too, but feel free to skip that!)
> 
> Once that's ready, we should get approval from @dkwingsmt as they worked in this space.

I noticed that `cursor_handler.cc` does not have unittest now. btw, can I manually generate a dummy cursor (buffer are all zero) for this test? I think a pure binary cursor is not suitable for the unittest.

Kingtous commented:
> 



> @Kingtous Thanks for following up quickly! Could you add a unit test that sends a `setSystemCursor` message and uses `MockWindowBindingHandler` to verify the cursor is set as expected? (If you want to go the extra mile, ideally `activateSystemCursor` would be tested too, but feel free to skip that!)
> 
> Once that's ready, we should get approval from @dkwingsmt as they worked in this space.

I've extracted the cursor creation logic to a function named `GetCursorFromBuffer`. And I added a unittest which will check if the `HCURSOR` is not null. I think this approach may keep this unit tests simple?

dkwingsmt commented on shell/platform/windows/cursor_handler.cc:
```diff
@@ -11,9 +11,22 @@
-
 static constexpr char kKindKey[] = "kind";
 
+// [kSetCustomCursorMethod] allows developers to set a custom cursor with
+// rawRGBA buffer.
```
```suggestion
// This method allows setting a custom cursor with rawRGBA buffer.
```

dkwingsmt commented on shell/platform/windows/cursor_handler.cc:
```diff
@@ -11,9 +11,22 @@
+// [kSetCustomCursorMethod] allows developers to set a custom cursor with
+// rawRGBA buffer.
+static constexpr char kSetCustomCursorMethod[] = "setCustomCursor";
+// std::vector<uint8_t> value for custom cursor rawRGBA buffer.
```
```suggestion
// A list of bytes, the custom cursor's rawRGBA buffer.
```

dkwingsmt commented on shell/platform/windows/cursor_handler.cc:
```diff
@@ -11,9 +11,22 @@
 
+// [kSetCustomCursorMethod] allows developers to set a custom cursor with
+// rawRGBA buffer.
+static constexpr char kSetCustomCursorMethod[] = "setCustomCursor";
```
Let's change it to "setCustomCursor/windows" to avoid confliction in the future.

dkwingsmt commented on shell/platform/windows/cursor_handler.cc:
```diff
@@ -11,9 +11,22 @@
+static constexpr char kSetCustomCursorMethod[] = "setCustomCursor";
+// std::vector<uint8_t> value for custom cursor rawRGBA buffer.
+static constexpr char kCustomCursorBufferKey[] = "buffer";
+// double value for the hotx value of custom cursor.
```
Change it to
```
// A double, the x coordinate of the custom cursor's hotspot, starting from left.
```
(Is the hotspot's origin at its top left?)

dkwingsmt commented on shell/platform/windows/cursor_handler.cc:
```diff
@@ -11,9 +11,22 @@
+static constexpr char kCustomCursorBufferKey[] = "buffer";
+// double value for the hotx value of custom cursor.
+static constexpr char kCustomCursorHotxKey[] = "hotx";
+// double value for the hoty value of custom cursor.
```
```
// A double, the y coordinate of the custom cursor's hotspot, starting from top.
```

dkwingsmt commented on shell/platform/windows/cursor_handler.cc:
```diff
@@ -11,9 +11,22 @@
+// std::vector<uint8_t> value for custom cursor rawRGBA buffer.
+static constexpr char kCustomCursorBufferKey[] = "buffer";
+// double value for the hotx value of custom cursor.
+static constexpr char kCustomCursorHotxKey[] = "hotx";
```
```suggestion
static constexpr char kCustomCursorHotXKey[] = "hotX";
```
Use camel case for key names.

dkwingsmt commented on shell/platform/windows/cursor_handler.cc:
```diff
@@ -11,9 +11,22 @@
+// double value for the hotx value of custom cursor.
+static constexpr char kCustomCursorHotxKey[] = "hotx";
+// double value for the hoty value of custom cursor.
+static constexpr char kCustomCursorHotyKey[] = "hoty";
```
```suggestion
static constexpr char kCustomCursorHotYKey[] = "hotY";
```

dkwingsmt commented on shell/platform/windows/cursor_handler.cc:
```diff
@@ -45,9 +58,86 @@ void CursorHandler::HandleMethodCall(
+                            int width,
+                            int height) {
+  HCURSOR cursor = nullptr;
+  // Flutter returns rawRgba, which has 8bits * 4channels
```
```suggestion
  // Flutter returns rawRgba, which has 8bits * 4channels.
```

dkwingsmt commented on shell/platform/windows/cursor_handler.h:
```diff
@@ -32,6 +32,13 @@ class CursorHandler {
   WindowBindingHandler* delegate_;
 };
 
+// Create cursor from buffer and cursor info
```
```suggestion
// Create a cursor from buffer and cursor info.
```

dkwingsmt commented on shell/platform/windows/cursor_handler.h:
```diff
@@ -32,6 +32,13 @@ class CursorHandler {
 
+// Create cursor from buffer and cursor info
+HCURSOR GetCursorFromBuffer(const std::vector<uint8_t>& buffer,
+                            double hotx,
+                            double hoty,
```
```suggestion
                            double hot_x,
                            double hot_y,
```
Among other places.

dkwingsmt commented on shell/platform/windows/cursor_handler_unittests.cc:
```diff
@@ -0,0 +1,27 @@
+namespace flutter {
+namespace testing {
+TEST(CursorHandlerTest, CreateDummyCursor) {
+  // create 4x4 rawRGBA dummy cursor buffer
```
```suggestion
  // Create 4x4 rawRGBA dummy cursor buffer.
```
Comments should be full sentences, starting with a capital letter and ending with a `.`. Among other places.

dkwingsmt commented:
@stuartmorgan What do you think?

Kingtous commented:
Thanks for the careful review. All suggestions have resolved :)

stuartmorgan commented:
> @stuartmorgan What do you think? (Also I'm not familiar enough with the capability of plugins.)

About what part in particular?

If the goal is to just land something that would allow a plugin to do their own cursor handling, the best approach would be to add a Flutter-view-level version of the top-level window proc delegation (see https://github.com/flutter/flutter/issues/53168#issuecomment-676531731, the only reason I didn't do it is that we didn't have a use case any more). It should be easy—it would look just like [this](https://github.com/flutter/engine/blob/main/shell/platform/windows/client_wrapper/include/flutter/plugin_registrar_windows.h#L52-L82), but for the Flutter view HWND instead of the top-level HWND—and it's a general-purpose API that may well have other uses in the future. If that were added, a plugin could just handled `WM_SETCURSOR` to do their own cursor management.

Kingtous commented:
> > @stuartmorgan What do you think? (Also I'm not familiar enough with the capability of plugins.)
> 
> About what part in particular?
> 
> If the goal is to just land something that would allow a plugin to do their own cursor handling, the best approach would be to add a Flutter-view-level version of the top-level window proc delegation (see [flutter/flutter#53168 (comment)](https://github.com/flutter/flutter/issues/53168#issuecomment-676531731), the only reason I didn't do it is that we didn't have a use case any more). It should be easy—it would look just like [this](https://github.com/flutter/engine/blob/main/shell/platform/windows/client_wrapper/include/flutter/plugin_registrar_windows.h#L52-L82), but for the Flutter view HWND instead of the top-level HWND—and it's a general-purpose API that may well have other uses in the future. If that were added, a plugin could just handle `WM_SETCURSOR` to do their own cursor management.

Yes, the problems is due to the issue: Currently plugins cannot get `WM_SETCURSOR` message from flutter because of this. 
https://github.com/flutter/engine/blob/243019f734a29acf13c87dfbf9dc12ee03a2d54d/shell/platform/windows/window.cc#L400-L408.

Flutter intercepts `WM_SETCURSOR` message and do `OnSetCursor`, then return `TRUE` to prevent message transfer down among plugins. Then the cursor will change to the cursor which created by flutter engine before.

We tried to handle `WM_SETCURSOR` in the plugins, but the handler never reached. 
https://github.com/Kingtous/rustdesk_flutter_custom_cursor/blob/f0c8f4cdbcf50fa9224bc4780c63fd577ba47f15/windows/flutter_custom_cursor_plugin.cpp#L72

Here is some clues:
Now, `OnSetCursor` methods cannot be removed here, because it must be here to keep cursor behavior normally. IF WE RETURN `FALSE` here, although plugins can handle this type of message, the cursor will not consistent:

- set custom cursor in plugin(to simplify, we call the custom cursor `CC`, now cursor is `CC`)
- flutter get `WM_SETCURSOR`, call `OnSetCursor`, cursor will change to default `Arrow` cursor. the `WM_SETCURSOR` continues to transfer down.(now cursor is `Arrow`)
- set custom cursor in plugin again. (`Arrow`-> `CC`)

So the cursor will change to arrow between two frames.

Another problem caused this issue is that Windows need to set cursor twice to successfully get specific cursor work, which is different from macOS/Linux. (So our plugins can handle custom cursors in macOS/Linux greatly)

To solve this problem gracefully, I think the best way is adding a method channel to change `cursor_` in flutter engine(which will be used to set cursor on `OnSetCursor`). So we can reuse current logic to keep custom cursor working, and keep engine simple.

stuartmorgan commented:
> Flutter intercepts `WM_SETCURSOR` message and do `OnSetCursor`, then return `TRUE` to prevent message transfer down among plugins. 

Flutter isn't "intercepting" it, it's handling it at the level of the Flutter view, which is the right place for `WM_SETCURSOR` to be handled.

> We tried to handle `WM_SETCURSOR` in the plugins, but the handler never reaches.

That's expected, because you're setting a handler for the **top-level** HWND, not the Flutter view HWND. Which is why my comment above is saying that the general fix would be to add plugin API to allow registering Flutter view HWND handlers in addition to top-level HWND handlers. Such a handler would be called before the engine code you linked to.

> Here are some clues why let plugins to handle `WM_SETCURSOR` is not a good way: Now, `OnSetCursor` methods cannot be removed here, because it must be here to keep cursor behavior normally.

My suggestion does not involve removing `OnSetCursor`.

> IF WE RETURN `FALSE` here, although plugins can handle this type of message, the cursor will not consistent.

My suggestion does not involve changing the return value of the existing code.

> Assuming we return `FALSE`, when we set a custom cursor in plugins:
> 
> * set custom cursor in plugin(to simplify, we call the custom cursor `CC`, now cursor is `CC`)
> * flutter get `WM_SETCURSOR`, call `OnSetCursor`, cursor will change to default `Arrow` cursor. the `WM_SETCURSOR` continues to transfer down.(now cursor is `Arrow`)
> * the plugin get `WM_SETCURSOR`, set custom cursor in plugin again. (`Arrow`-> `CC`)

That is not the flow that would happen with my suggested solution. The flow would be:
- The plugin gets `WM_SETCURSOR` (via the handler it registers with the new `RegisterFlutterViewWindowProcDelegate`, or whatever we call it), sets the cursor to `CC`, and returns `TRUE` because it handled it.
- The engine's `WM_SETCURSOR` is never called because the plugin code, which ran first, returned `TRUE`.

> To solve this problem gracefully, I think the best way is adding a method channel to change `cursor_` in flutter engine(which will be used to set cursor on `OnSetCursor`). So we can reuse current logic to keep custom cursor working, and keep engine simple and reuseable.

I can't speak to whether adding the method channel code proposed here is a good idea; @dkwingsmt raised concerns with it above, and knows far more about the future plans here than I do. But if @dkwingsmt would prefer to enable plugin-level solutions for now instead, it could be done very easily, by adding a general-purpose API that I had always intended to add if there was ever a use case for it.

Kingtous commented:
Thanks for the reply!

> Here are some clues why let plugins to handle WM_SETCURSOR is not a good way: Now, OnSetCursor methods cannot be removed here, because it must be here to keep cursor behavior normally.

It's is just my inspiration for adding a method channel to change `cursor_` field, not considering to let `WM_SETCURSOR` return `TRUE` :).

> The plugin gets WM_SETCURSOR (via the handler it registers with the new RegisterFlutterViewWindowProcDelegate, or whatever we call it), sets the cursor to CC, and returns TRUE because it handled it.
The engine's WM_SETCURSOR handling is never called because the plugin code, which ran first, returned TRUE.

Sounds like a more generic solution to solve this problem than I proposed. I dont know which is a better solution for it. As u mentioned, there are few use cases to register a `FlutterView` level window delegate. 

> However, I'm hesitant on merging the method channel, since we are hoping to design a universal API (method channel) for all platforms, and the current method is likely platform dependent. But I also understand that we need to develop one step at a time. So I think there are two options:

I think @dkwingsmt may want to integrate cursor logic in the engine and develops a universal custom cursor API for all platforms (an engine-level solution for customizing cursors?). 

btw, if needed to add this `RegisterFlutterViewWindowProcDelegate` or revise this pr, I am willing to help :)

Kingtous commented:
Any good news here about this issue?

I've split the `setCustomCursor` channel be 3 ones (`createCustomCursor`, `deleteCustomCursor`, `setCustomCursor`), just like @dkwingsmt mentioned here, which gives a better way to handle customizing cursors.

> Also, the flow I'm imagining is, instead of setting the cursor using the buffer directly, which you have to do every time you set > the same cursor,

> Create a cursor using the buffer, and assign it with a name (string)
> Activate the cursor using the string.

- `createCustomCursor`: used to create a custom cursor, if success, a unique index(key) is returned with `result->Success`
- `deleteCustomCursor`: used to delete a custom cursor from the cache with the provided index.
- `setCustomCursor`: used to set a custom cursor from the cache.

The reason I didn't choose to use string for keys is: int64_t index comparation is more efficient than string ones, and it can be used to get `HCURSOR` directly by an offset.

dkwingsmt commented on shell/platform/windows/cursor_handler.cc:
```diff
@@ -45,9 +70,180 @@ void CursorHandler::HandleMethodCall(
+    }
+    // push the cursor into the cache vector of this handler.
+    custom_cursors_.emplace_back(std::move(cursor));
+    int64_t key = custom_cursors_.size() - 1;
```
We should probably not derive `key` from the size of the vector `custom_cursors_`. If a key is ever removed, then all following keys are invalidated, but the user (framework code) does not know it.

Kingtous replied:
I've replaced `vector` with `unordered_map`, which accepts a `<std::string, HCURSOR>` pair.

dkwingsmt commented:
My apology for being busy for a while. The PR looks generally OK. I think the method channel is a good enough implementation for now since we're bound to make the universal API compatible anyway.

However, I'm not sure about using vector indexes as cursor ID (see comment). I suggest letting the framework provide the ID in the method call argument, instead of generating it from the engine and return it.

Kingtous commented:
> My apology for being late. The PR looks generally OK. I think the method channel is a good enough implementation for now since we're bound to make the universal API compatible anyway.
> 
> However, I'm not sure about using vector indexes as cursor ID (see comment). I suggest letting the framework provide the ID in the method call argument (instead of generating it from the engine and return it) and mapping the ID to the cursors in a map.

Thanks for the kind reply and careful review. I will prepare the refactor for providing key from the framework as soon as possible.

dkwingsmt commented on shell/platform/windows/cursor_handler.cc:
```diff
@@ -11,9 +11,36 @@
+// custom cursor.
+static constexpr char kSetCustomCursorMethod[] = "setCustomCursor/windows";
+// A string value for the key of the custom cursor.
+static constexpr char kCustomCursorKey[] = "key";
```
Why is this `key`? Shouldn't it just use `kCustomCursorNameKey`?

Kingtous replied:
my apology that I forget to uniform this key. Now it's has been resolved.

dkwingsmt commented on shell/platform/windows/cursor_handler.cc:
```diff
@@ -11,9 +11,34 @@
+static constexpr char kCustomCursorHeightKey[] = "height";
+
+// This method allows setting a custom cursor with a unique int64_t key of the
+// custom cursor.
```
```suggestion
// custom cursor.
//
// This method also has an argument `kCustomCursorNameKey` for the name
// of the cursor to activate.
```

dkwingsmt commented on shell/platform/windows/cursor_handler.cc:
```diff
@@ -11,9 +11,34 @@
+// custom cursor.
+static constexpr char kSetCustomCursorMethod[] = "setCustomCursor/windows";
+
+// This method allows deleting a custom cursor with a string key.
```
```suggestion
// This method allows deleting a custom cursor with a string key.
//
// This method also has an argument `kCustomCursorNameKey` for the name
// of the cursor to delete.
```

dkwingsmt commented on shell/platform/windows/cursor_handler.cc:
```diff
@@ -45,9 +70,192 @@ void CursorHandler::HandleMethodCall(
+                    "Argument must contains a valid rawBGRA bitmap");
+      return;
+    }
+    // push the cursor into the cache vector of this handler.
```
```suggestion
    // Push the cursor into the cache map.
```

dkwingsmt commented on shell/platform/windows/cursor_handler.cc:
```diff
@@ -45,9 +70,192 @@ void CursorHandler::HandleMethodCall(
+    }
+    auto key = std::get<std::string>(key_iter->second);
+    auto it = custom_cursors_.find(key);
+    // if the cursor identified by key cannot be found, it's ok for deleting
+    // operations
```
```suggestion
    // If the specified cursor name is not found, the deletion is a noop and returns success.
```

dkwingsmt commented on shell/platform/windows/cursor_handler_unittests.cc:
```diff
@@ -0,0 +1,27 @@
+namespace flutter {
+namespace testing {
+TEST(CursorHandlerTest, CreateDummyCursor) {
+  // create a 4x4 rawBGRA dummy cursor buffer.
```
```suggestion
  // Create a 4x4 rawBGRA dummy cursor buffer.
```

dkwingsmt commented on shell/platform/windows/cursor_handler_unittests.cc:
```diff
@@ -0,0 +1,27 @@
+  for (int i = 0; i < 4 * 4 * 4; i++) {
+    buffer.push_back(0);
+  }
+  // create the cursor from buffer provided above.
```
```suggestion
  // Create the cursor from buffer provided above.
```

dkwingsmt commented on shell/platform/windows/cursor_handler_unittests.cc:
```diff
@@ -0,0 +1,27 @@
+  }
+  // create the cursor from buffer provided above.
+  auto cursor = GetCursorFromBuffer(buffer, 0.0, 0.0, 4, 4);
+  // expect cursor is not null.
```
```suggestion
  // Expect cursor is not null.
```

dkwingsmt commented on shell/platform/windows/window_binding_handler.h:
```diff
@@ -72,6 +72,9 @@ class WindowBindingHandler {
   // content. See mouse_cursor.dart for the values and meanings of cursor_name.
   virtual void UpdateFlutterCursor(const std::string& cursor_name) = 0;
 
+  // Sets the cursor directly from cursor
```
```suggestion
  // Sets the cursor directly from a cursor handle.
```

dkwingsmt reviewed (approved):
LGTM except for comment details.

Kingtous commented:
Appreciated. To ensure it's worked with the latest master channel of flutter. I'll test the commit with our custom cursor plugin again. Sorry for a little delay :)

dkwingsmt commented:
No problem, let me know when it's ready :) Thank you for working on this.

Kingtous commented:
> No problem, let me know when it's ready :) Thank you so much for working on this!

I've tested with the latest flutter/engine, it works perfectly with https://github.com/Kingtous/rustdesk_flutter_custom_cursor/blob/refactor/flutter/master/lib/cursor_manager.dart

The test cursor has successfully shown without a delay or an error.

dkwingsmt commented:
@stuartmorgan Would you like to give a second review?

dkwingsmt commented:
@Kingtous I've approved this PR, but it is required that each PR needs the participance of 2 members of flutter/flutter-hackers, or in the case of this PR, 2 approvals. I'm looking for a second reviewer.

SahajRana commented:
How long will this take?

chinmaygarde commented:
@dkwingsmt This looks good to go. Please CQ when ready :)
